backend/controller/evaljob_controller.go

```diff
@@ -0,0 +1,568 @@
+			Name:      job.Status.PodName,
+			Namespace: job.Namespace,
+			OwnerReferences: []v1.OwnerReference{
+				{
+					APIVersion: job.APIVersion,
+					Kind:       job.Kind,
+					Name:       job.Name,
+				},
+			},
+		},
+	}
+	return r.Delete(ctx, &pod, &client.DeleteOptions{})
+}
+
+func (r *EvalJobReconciler) handleComplete(ctx context.Context, log logr.Logger, job *lmevalservicev1beta1.EvalJob) (ctrl.Result, error) {
+	if job.Status.CompleteTime == nil {
+		r.Recorder.Event(job, "Normal", "JobCompleted",
+			fmt.Sprintf("Tthe EvalJob %s in namespace %s has completed",
+				job.Name,
+				job.Namespace))
+		// TODO: final wrap up/clean up
+		current := v1.Now()
+		job.Status.CompleteTime = &current
+		if err := r.Status().Update(ctx, job); err != nil {
+			log.Error(err, "failed to update status for completion")
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *EvalJobReconciler) handleCancel(ctx context.Context, log logr.Logger, job *lmevalservicev1beta1.EvalJob) (ctrl.Result, error) {
+	// delete the pod and update the state to complete
+	if _, err := r.getPod(ctx, job); err != nil {
+		// pod is gone. update status
+		job.Status.State = lmevalservicev1beta1.CompleteJobState
+		job.Status.Reason = lmevalservicev1beta1.FailedReason
+		job.Status.Message = err.Error()
+	} else {
+		job.Status.State = lmevalservicev1beta1.CompleteJobState
+		job.Status.Reason = lmevalservicev1beta1.CancelledReason
+		if err := r.deleteJobPod(ctx, job); err != nil {
+			// leave the state as is and retry again
+			log.Error(err, "failed to delete pod. scheduled a retry after 10 seconds")
+			return ctrl.Result{Requeue: true, RequeueAfter: time.Second * 10}, err
+		}
+	}
+
+	err := r.Status().Update(ctx, job)
+	if err != nil {
+		log.Error(err, "failed to update status for cancellation")
+	}
+	r.Recorder.Event(job, "Normal", "Cancelled",
+		fmt.Sprintf("Tthe EvalJob %s in namespace %s has cancelled and changed its state to Complete",
+			job.Name,
+			job.Namespace))
+	return ctrl.Result{}, err
+}
+
+func (r *EvalJobReconciler) createPod(job *lmevalservicev1beta1.EvalJob) *corev1.Pod {
+	var allowPrivilegeEscalation = false
+	var runAsNonRootUser = true
+	var ownerRefController = true
+	var runAsUser int64 = 1001030000
+
+	// Then compose the Pod CR
+	pod := corev1.Pod{
+		TypeMeta: v1.TypeMeta{
+			Kind:       "Pod",
+			APIVersion: "v1",
+		},
+		ObjectMeta: v1.ObjectMeta{
+			Name:      job.Name,
+			Namespace: job.Namespace,
+			OwnerReferences: []v1.OwnerReference{
+				{
+					APIVersion: job.APIVersion,
+					Kind:       job.Kind,
+					Name:       job.Name,
+					Controller: &ownerRefController,
+					UID:        job.UID,
+				},
+			},
+			Labels: map[string]string{
+				"app.kubernetes.io/name": "fms-lm-eval-service",
+			},
+		},
+		Spec: corev1.PodSpec{
+			InitContainers: []corev1.Container{
+				{
+					Name:            "driver",
+					Image:           r.options.DriverImage,
+					ImagePullPolicy: corev1.PullAlways,
```

Comment:
I think this will probably need to be configurable too. I always struggle with this kind of `controller` on how much to make the CRD's `spec` include a generic "put anything to configure the pod here" section vs being prescriptive about the body of the pod.

Reply:
Since this is used by the controller and we do have a configmap for the settings. I can add the image pull policy there.